### ISSUE TITLE
Wrong error message when using `Self` in some places

### ISSUE THREAD

**Issue by bombless** [rust-lang/rust#24968]
```rust
fn foo(_: Self) {}

fn main() {}
```
```
<anon>:1:11: 1:15 error: use of undeclared type name `Self`
<anon>:1 fn foo(_: Self) {}
                   ^~~~
```

`Self` is a keyword, this error message indicates that it can be used to define type, which is misleading.

**Comment by XuefengWu** [comment]
There is a similar issue: https://github.com/rust-lang/rust/issues/12796 file.

**Comment by bombless** [comment]
No, that's a different issue.
At the time that issue was filed, `Self` was not a keyword, just a confusing type. That said, at the time that issue was reported, it's fine to say the compiler was implemented properly.
But my issue is about wrong error message, not bad(confusing) error message.

**Comment by XuefengWu** [comment]
Do you mean `Self` is not a special ident now?
I found the example code can be compiled,
```
pub trait Square {
    fn square(x: Self) -> Self;
}
```

and there is excluding self type code。
```
pub fn parse_ident_or_self_type(&mut self) -> PResult<ast::Ident> {
        if self.is_self_type_ident() {
            self.expect_self_type_ident()
        } else {
            self.parse_ident()
        }
    }
```
https://github.com/XuefengWu/rust/blob/master/src/libsyntax/parse/parser.rs#L558-L564

**Comment by bombless** [comment]
My point is that we are in different situation now.
See https://github.com/rust-lang/rust/pull/22158 and notice that it happened after the time the issue we discuss was filed.

**Comment by Manishearth** [comment]
Actually, IIRC we're moving towards a direction where `Self` can be used everywhere in a trait or impl context. I'm not sure though. @pnkfelix any idea?

**Comment by bombless** [comment]
Yes we can use it in traits or impls.
```rust
trait T where Self: Sized {}
```
is totally legal for instance.

But `Self` inside global functions can not be explained if I understand correctly.

**Comment by Manishearth** [comment]
Oh, sorry, this was a global fn. Makes sense.
